Predictive modeling of NHANES Data using Python
Predictive modeling workflows using classification and regression models. This project demonstrates this process using publicly available health datasets from the CDC's [2017-2018 National Health and Nutrition Examination Survey](https://wwwn.cdc.gov/nchs/nhanes/continuousnhanes/default.aspx?BeginYear=2017) (NHANES).



# Part 1: Classification modeling

# Step 1: Retrieve data
The code below uses Python libraries to:

*   Download three datasets directly from the CDC website: demographics (DEMO_J), physical activity (PAQ_J), body measurements (BMX_J), and diabetes (DIQ_J).
*   Load the datasets using pandas and the read_sas() function, which supports SAS transport files (.XPT format).
*   Display the first five rows of each dataset to explore their structure and contents.





In [ ]:
# Import needed libraries
import pandas as pd
import urllib.request

# Download files directly from CDC NHANES site
urllib.request.urlretrieve("https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/DEMO_J.xpt", "DEMO_J.XPT")
urllib.request.urlretrieve("https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/PAQ_J.xpt", "PAQ_J.XPT")
urllib.request.urlretrieve("https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/BMX_J.xpt", "BMX_J.XPT")
urllib.request.urlretrieve("https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/DIQ_J.xpt", "DIQ_J.XPT")

# Load the datasets
demo = pd.read_sas("DEMO_J.XPT")
paq = pd.read_sas("PAQ_J.XPT")
bmx = pd.read_sas("BMX_J.XPT")
diq = pd.read_sas("DIQ_J.XPT")

# Preview the data
print("Demographics:")
display(demo.head())

print("Physical Activity:")
display(paq.head())

print("Body Measures:")
display(bmx.head())

# Preview the data
print("Diabetes:")
display(diq.head())

Demographics:


,SEQN,SDDSRVYR,RIDSTATR,RIAGENDR,RIDAGEYR,RIDAGEMN,RIDRETH1,RIDRETH3,RIDEXMON,RIDEXAGM,...,DMDHREDZ,DMDHRMAZ,DMDHSEDZ,WTINT2YR,WTMEC2YR,SDMVPSU,SDMVSTRA,INDHHIN2,INDFMIN2,INDFMPIR
0,93703.0,10.0,2.0,2.0,2.0,NaN,5.0,6.0,2.0,27.0,...,3.0,1.0,3.0,9246.491865,8539.731348,2.0,145.0,15.0,15.0,5.00
1,93704.0,10.0,2.0,1.0,2.0,NaN,3.0,3.0,1.0,33.0,...,3.0,1.0,2.0,37338.768343,42566.614750,1.0,143.0,15.0,15.0,5.00
2,93705.0,10.0,2.0,2.0,66.0,NaN,4.0,4.0,2.0,NaN,...,1.0,2.0,NaN,8614.571172,8338.419786,2.0,145.0,3.0,3.0,0.82
3,93706.0,10.0,2.0,1.0,18.0,NaN,5.0,6.0,2.0,222.0,...,3.0,1.0,2.0,8548.632619,8723.439814,2.0,134.0,NaN,NaN,NaN
4,93707.0,10.0,2.0,1.0,13.0,NaN,5.0,7.0,2.0,158.0,...,2.0,1.0,3.0,6769.344567,7064.609730,1.0,138.0,10.0,10.0,1.88


Physical Activity:


,SEQN,PAQ605,PAQ610,PAD615,PAQ620,PAQ625,PAD630,PAQ635,PAQ640,PAD645,PAQ650,PAQ655,PAD660,PAQ665,PAQ670,PAD675,PAD680
0,93705.0,2.0,NaN,NaN,2.0,NaN,NaN,2.0,NaN,NaN,2.0,NaN,NaN,1.0,2.0,60.0,300.0
1,93706.0,2.0,NaN,NaN,2.0,NaN,NaN,1.0,5.0,45.0,2.0,NaN,NaN,1.0,2.0,30.0,240.0
2,93708.0,2.0,NaN,NaN,2.0,NaN,NaN,2.0,NaN,NaN,2.0,NaN,NaN,1.0,5.0,30.0,120.0
3,93709.0,2.0,NaN,NaN,1.0,2.0,180.0,2.0,NaN,NaN,2.0,NaN,NaN,2.0,NaN,NaN,600.0
4,93711.0,2.0,NaN,NaN,2.0,NaN,NaN,1.0,5.0,60.0,1.0,4.0,60.0,1.0,2.0,30.0,420.0


Body Measures:


,SEQN,BMDSTATS,BMXWT,BMIWT,BMXRECUM,BMIRECUM,BMXHEAD,BMIHEAD,BMXHT,BMIHT,...,BMXLEG,BMILEG,BMXARML,BMIARML,BMXARMC,BMIARMC,BMXWAIST,BMIWAIST,BMXHIP,BMIHIP
0,93703.0,1.0,13.7,3.0,89.6,NaN,NaN,NaN,88.6,NaN,...,NaN,NaN,18.0,NaN,16.2,NaN,48.2,NaN,NaN,NaN
1,93704.0,1.0,13.9,NaN,95.0,NaN,NaN,NaN,94.2,NaN,...,NaN,NaN,18.6,NaN,15.2,NaN,50.0,NaN,NaN,NaN
2,93705.0,1.0,79.5,NaN,NaN,NaN,NaN,NaN,158.3,NaN,...,37.0,NaN,36.0,NaN,32.0,NaN,101.8,NaN,110.0,NaN
3,93706.0,1.0,66.3,NaN,NaN,NaN,NaN,NaN,175.7,NaN,...,46.6,NaN,38.8,NaN,27.0,NaN,79.3,NaN,94.4,NaN
4,93707.0,1.0,45.4,NaN,NaN,NaN,NaN,NaN,158.4,NaN,...,38.1,NaN,33.8,NaN,21.5,NaN,64.1,NaN,83.0,NaN


Diabetes:


,SEQN,DIQ010,DID040,DIQ160,DIQ170,DIQ172,DIQ175A,DIQ175B,DIQ175C,DIQ175D,...,DIQ300D,DID310S,DID310D,DID320,DID330,DID341,DID350,DIQ350U,DIQ360,DIQ080
0,93703.0,2.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,93704.0,2.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,93705.0,2.0,NaN,2.0,2.0,2.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,93706.0,2.0,NaN,2.0,2.0,2.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,93707.0,2.0,NaN,2.0,2.0,2.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


# Step 2: Merge data

In [ ]:
# Merge datasets on SEQN
merged = demo.merge(paq, on='SEQN').merge(bmx, on='SEQN').merge(diq, on='SEQN')

# Preview the merged dataset
print("Merged Dataset:")
display(merged.head())

Merged Dataset:


,SEQN,SDDSRVYR,RIDSTATR,RIAGENDR,RIDAGEYR,RIDAGEMN,RIDRETH1,RIDRETH3,RIDEXMON,RIDEXAGM,...,DIQ300D,DID310S,DID310D,DID320,DID330,DID341,DID350,DIQ350U,DIQ360,DIQ080
0,93705.0,10.0,2.0,2.0,66.0,NaN,4.0,4.0,2.0,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,93706.0,10.0,2.0,1.0,18.0,NaN,5.0,6.0,2.0,222.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,93708.0,10.0,2.0,2.0,66.0,NaN,5.0,6.0,2.0,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,93709.0,10.0,2.0,2.0,75.0,NaN,4.0,4.0,1.0,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,93711.0,10.0,2.0,1.0,56.0,NaN,5.0,6.0,2.0,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


# Step 3: Select & rename relevant columns

In [ ]:
# Select a subset of relevant columns
selected = merged[['RIDAGEYR', 'RIAGENDR', 'DMDEDUC2', 'BMXBMI', 'PAD660', 'DIQ010']]

# Rename columns for clarity
selected.columns = ['Age', 'Gender', 'Education', 'BMI', 'PhysicalActivity', 'DiabetesDiagnosed']

# Step 4: Handle missing values

In [ ]:
# Check for missing values
print(selected.isnull().sum())

Age                     0
Gender                  0
Education             268
BMI                    99
PhysicalActivity     4183
DiabetesDiagnosed       0
dtype: int64


In [ ]:
# Drop missing values
selected = selected.dropna()

# Preview cleaned data
print("Number of rows:", len(selected))
print("Cleaned Dataset:")
display(selected.head())

Number of rows: 1206
Cleaned Dataset:


,Age,Gender,Education,BMI,PhysicalActivity,DiabetesDiagnosed
4,56.0,1.0,5.0,21.3,60.0,2.0
8,71.0,1.0,3.0,22.5,60.0,2.0
9,61.0,1.0,5.0,30.7,180.0,2.0
14,64.0,1.0,5.0,22.4,60.0,3.0
19,57.0,1.0,2.0,28.6,120.0,1.0


# Step 5: Prepare target variable for classification

In [ ]:
# Determine what the unique values are used for Diabetes
print(selected['DiabetesDiagnosed'].unique())

[2. 3. 1. 9.]


In [ ]:

# Convert to numeric if needed
selected['DiabetesDiagnosed'] = pd.to_numeric(selected['DiabetesDiagnosed'], errors='coerce')

# Keep only valid responses (1 = Yes, 2 = No)
selected = selected[selected['DiabetesDiagnosed'].isin([1, 2])]

# Recode to binary: 1 = Diabetes, 0 = No Diabetes
selected['DiabetesDiagnosed'] = selected['DiabetesDiagnosed'].replace({1: 1, 2: 0})

# Check distribution
print("DiabetesDiagnosed value counts:")
print(selected['DiabetesDiagnosed'].value_counts())

print("Number of rows:", len(selected))

DiabetesDiagnosed value counts:
DiabetesDiagnosed
0.0    1096
1.0      78
Name: count, dtype: int64
Number of rows: 1174


In [ ]:
# Rename the dataset to selected_clean
selected_clean=selected

print("Cleaned Dataset:")
display(selected_clean.head(25))

Cleaned Dataset:


,Age,Gender,Education,BMI,PhysicalActivity,DiabetesDiagnosed
4,56.0,1.0,5.0,21.3,60.0,0.0
8,71.0,1.0,3.0,22.5,60.0,0.0
9,61.0,1.0,5.0,30.7,180.0,0.0
19,57.0,1.0,2.0,28.6,120.0,1.0
20,20.0,1.0,4.0,27.6,120.0,0.0
23,26.0,1.0,4.0,33.7,60.0,0.0
30,26.0,2.0,4.0,29.4,60.0,0.0
34,60.0,1.0,5.0,29.1,30.0,1.0
40,36.0,2.0,2.0,27.1,120.0,1.0
47,47.0,2.0,4.0,25.9,60.0,0.0


# Step 6: Define predictors and target
*   X contains the predictor variables (features) used to train the model.
*   y is the target variable—whether the participant has been diagnosed with diabetes (binary: 0 or 1).


In [ ]:
# Define predictors and target
X = selected_clean[['Age', 'Gender', 'Education', 'BMI', 'PhysicalActivity']]
y = selected_clean['DiabetesDiagnosed']

# Step 7: Train-test split
*   80% for training the model (X_train, y_train)
*   20% for testing the model (X_test, y_test)
random_state=42 ensures reproducibility.


In [ ]:
# Train-test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 8: Classification: Fit logistic regression model
*   A logistic regression model is trained to predict the probability of diabetes diagnosis.
*   max_iter=1000 increases the number of iterations to ensure convergence.

In [ ]:
# Fit logistic regression model
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

# Predict and evaluate
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))


Accuracy: 0.9276595744680851
Confusion Matrix:
 [[217   0]
 [ 17   1]]


# Interpretation
*  92.77% of the predictions made by the model were correct.
*  True Negatives (217): The model correctly predicted "No" for participants who were not diagnosed with diabetes.
*   True Positives (1): The model correctly predicted "Yes" for one participant who was diagnosed with diabetes.
*   False Negatives (17): The model incorrectly predicted "No" for 17 participants who actually diagnosed with diabetes.
*   False Positives (0): The model did not incorrectly predict "Yes" for any participant who was not diagnosed with diabetes.
*   Precision = 1 / (1 + 0) = 1.0
*   Recall = 1 / (1 + 17) = 0.0556
*   High precision but very low recall—the model rarely identifies true positives.

This model is very good at identifying people without diabetes, but struggles to identify those with diabetes. This is likely due to class imbalance—far more people in the dataset do not have diabetes.

# Part 2: Regression modeling
*   Predicting BMI (Continuous Outcome)
*   Model type: Linear regression

*   Outcome: BMI
*   Predictors: Age, Gender, Education, PhysicalActivity, DiabetesDiagnosed

This model is useful if you want to understand how demographic and behavioral factors relate to body mass.

# Step 1: Import libraries
*   pandas as pd: Imports the pandas library for data manipulation and analysis.
*   statsmodels.api as sm: Imports the statsmodels library, which is used for statistical modeling (e.g., linear regression).



In [ ]:
# Import libraries
import pandas as pd
import statsmodels.api as sm

*   X: This is the design matrix (independent variables or predictors).
*   y: This is the response variable (dependent variable), which is BMI in this case.
*   Adds a constant (intercept) term to the model. This is necessary because statsmodels.OLS does not automatically include an intercept. Without this, the model would be forced through the origin (0,0), which is usually not appropriate.

In [ ]:
# Define predictors and outcome
X = selected_clean[['Age', 'Gender', 'Education', 'PhysicalActivity', 'DiabetesDiagnosed']]
y = selected_clean['BMI']

# Add constant for intercept
X = sm.add_constant(X)

# Step 2: Check for multicollinearity

**What a variance inflation factor (VIF) tells you:**
*   VIF = 1: No correlation between the predictor and other variables.
*   VIF between 1 and 5: Moderate correlation, usually acceptable.
*   VIF > 5: Potential multicollinearity concern.
*   VIF > 10: Serious multicollinearity problem; consider removing or combining variables.

**Why it matters- High multicollinearity can:**
*   Inflate the standard errors of coefficients.
*   Make model estimates unstable and hard to interpret.
*   Reduce the reliability of statistical tests.

**Intepretation of results (below):**
All the predictors have a VIF < 2, which is excellent. This means:
*   There is no significant multicollinearity among your predictors.
*   We can proceed confidently with your regression model using these variables.


In [ ]:
import pandas as pd
from statsmodels.stats.outliers_influence import variance_inflation_factor
from patsy import dmatrices

# Create design matrices to handle categorical variables automatically
y, X = dmatrices('BMI ~ Age + Gender + Education + PhysicalActivity + DiabetesDiagnosed',
                 data=selected_clean, return_type='dataframe')

# Calculate VIF for each predictor
vif_data = pd.DataFrame()
vif_data['Feature'] = X.columns
vif_data['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

# Display the VIF values
print(vif_data)

             Feature        VIF
0          Intercept  27.121778
1                Age   1.083888
2             Gender   1.023727
3          Education   1.049918
4   PhysicalActivity   1.014733
5  DiabetesDiagnosed   1.087260


# Step 3: Fit linear regression model
*   sm.OLS(y, X): Specifies an Ordinary Least Squares (OLS) regression model with y as the outcome and X as the predictors.
*   .fit(): Fits the model to the data, estimating the regression coefficients.

In [ ]:
# Fit linear regression model
model = sm.OLS(y, X).fit()

# Display model summary
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:                    BMI   R-squared:                       0.027
Model:                            OLS   Adj. R-squared:                  0.023
Method:                 Least Squares   F-statistic:                     6.509
Date:                Mon, 30 Mar 2026   Prob (F-statistic):           5.57e-06
Time:                        10:59:52   Log-Likelihood:                -3834.0
No. Observations:                1174   AIC:                             7680.
Df Residuals:                    1168   BIC:                             7710.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept            28.8722      0.96

# Interpretation
*   R-squared = 0.027:
This means that only 2.7% of the variance in BMI is explained by your predictors. This is very low, suggesting that the model does not capture much of the variation in BMI.
*   Adjusted R-squared = 0.023:
Slightly lower than R-squared, which accounts for the number of predictors. This confirms that adding these variables did not improve the model much.
*   F-statistic = 6.509, p-value = 5.57e-06:
The overall model is statistically significant, meaning at least one predictor is likely contributing meaningfully. However, the practical significance is weak due to the low R-squared.

**Key takeaways:**
*   The model is statistically significant but not practically useful for predicting BMI.
*   DiabetesDiagnosed is the only meaningful predictor in this model.
*   We may need to:
    *   Add more relevant predictors (e.g., dietary intake, medication use, socioeconomic status).
    *   Consider non-linear models or interactions.
    *   Explore feature engineering or clustering to uncover subgroups.





# Customization

1. Replace the 'PhysicalActivity' variable with 'SedentaryTime' (found in PAQ_J) in both the classification and regression models.
2. Fit the updated models and examine the results.



In [ ]:
# Select a subset of relevant columns
selected = merged[['RIDAGEYR', 'RIAGENDR', 'DMDEDUC2', 'BMXBMI', 'PAD680', 'DIQ010']]

# Rename columns for clarity
selected.columns = ['Age', 'Gender', 'Education', 'BMI', 'SedentaryTime', 'DiabetesDiagnosed']

## Check Missing values

In [ ]:
# Check for missing values
print(selected.isnull().sum())

Age                    0
Gender                 0
Education            268
BMI                   99
SedentaryTime         10
DiabetesDiagnosed      0
dtype: int64


In [ ]:
# Drop missing values
selected = selected.dropna()

# Preview cleaned data
print("Number of rows:", len(selected))
print("Cleaned Dataset:")
display(selected.head())

Number of rows: 5166
Cleaned Dataset:


,Age,Gender,Education,BMI,SedentaryTime,DiabetesDiagnosed
0,66.0,2.0,2.0,31.7,300.0,2.0
2,66.0,2.0,1.0,23.7,120.0,3.0
3,75.0,2.0,4.0,38.9,600.0,2.0
4,56.0,1.0,5.0,21.3,420.0,2.0
6,67.0,1.0,3.0,23.5,120.0,2.0


## Target Variables Prep


In [ ]:
# Determine what the unique values are used for Diabetes
print(selected['DiabetesDiagnosed'].unique())

[2. 3. 1. 9.]


In [ ]:
# Convert to numeric if needed
selected['DiabetesDiagnosed'] = pd.to_numeric(selected['DiabetesDiagnosed'], errors='coerce')

# Keep only valid responses (1 = Yes, 2 = No)
selected = selected[selected['DiabetesDiagnosed'].isin([1, 2])]

# Recode to binary: 1 = Diabetes, 0 = No Diabetes
selected['DiabetesDiagnosed'] = selected['DiabetesDiagnosed'].replace({1: 1, 2: 0})

# Check distribution
print("DiabetesDiagnosed value counts:")
print(selected['DiabetesDiagnosed'].value_counts())

print("Number of rows:", len(selected))

DiabetesDiagnosed value counts:
DiabetesDiagnosed
0.0    4187
1.0     814
Name: count, dtype: int64
Number of rows: 5001


In [ ]:
# Rename the dataset to selected_clean
selected_clean=selected

print("Cleaned Dataset:")
display(selected_clean.head(25))

Cleaned Dataset:


,Age,Gender,Education,BMI,SedentaryTime,DiabetesDiagnosed
0,66.0,2.0,2.0,31.7,300.0,0.0
3,75.0,2.0,4.0,38.9,600.0,0.0
4,56.0,1.0,5.0,21.3,420.0,0.0
6,67.0,1.0,3.0,23.5,120.0,0.0
7,54.0,2.0,4.0,39.9,360.0,1.0
8,71.0,1.0,3.0,22.5,180.0,0.0
9,61.0,1.0,5.0,30.7,300.0,0.0
10,22.0,1.0,3.0,24.5,300.0,0.0
11,45.0,1.0,3.0,22.0,240.0,0.0
12,60.0,2.0,1.0,35.9,120.0,0.0


## Define predictors

In [ ]:
# Define predictors and target
X = selected_clean[['Age', 'Gender', 'Education', 'BMI', 'SedentaryTime']]
y = selected_clean['DiabetesDiagnosed']

## Train-test split

In [ ]:
# Train-test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Classification: fit logistic regression model

In [ ]:
# Fit logistic regression model
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

# Predict and evaluate
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.8251748251748252
Confusion Matrix:
 [[814  14]
 [161  12]]


## Interpretation
* Accuracy = 0.825 (82.5%)
* True Negatives (814) → correctly predicted No diabetes
* False Positives (14) → predicted diabetes, but actually no diabetes
* False Negatives (161) → predicted no diabetes, but actually has diabetes
* True Positives (12) → correctly predicted diabetes

The model using **physical activity achieved higher overall accuracy (92.77%) compared to the sedentary behavior model (82.5%)**, but this was misleading due to class imbalance. The physical activity model correctly identified only one diabetes case, resulting in extremely low recall (5.6%). In contrast, the sedentary behavior model detected more diabetes cases (12 true positives) and showed slightly improved sensitivity. Although both models performed poorly in identifying diabetes, sedentary behavior appeared to be a more informative predictor than physical activity. These findings suggest that relying on a single lifestyle variable is insufficient for diabetes prediction, and additional clinical and demographic factors are necessary to improve model performance.

## Regression modeling

In [ ]:
# Import libraries
import pandas as pd
import statsmodels.api as sm

In [ ]:
# Define predictors and outcome
X = selected_clean[['Age', 'Gender', 'Education', 'SedentaryTime', 'DiabetesDiagnosed']]
y = selected_clean['BMI']

# Add constant for intercept
X = sm.add_constant(X)

## Check for multicollinearity

In [ ]:
import pandas as pd
from statsmodels.stats.outliers_influence import variance_inflation_factor
from patsy import dmatrices

# Create design matrices to handle categorical variables automatically
y, X = dmatrices('BMI ~ Age + Gender + Education + SedentaryTime + DiabetesDiagnosed',
                 data=selected_clean, return_type='dataframe')

# Calculate VIF for each predictor
vif_data = pd.DataFrame()
vif_data['Feature'] = X.columns
vif_data['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

# Display the VIF values
print(vif_data)

             Feature        VIF
0          Intercept  28.109417
1                Age   1.111741
2             Gender   1.005506
3          Education   1.013676
4      SedentaryTime   1.004115
5  DiabetesDiagnosed   1.115997


## Fit linear regression model

In [ ]:
# Fit linear regression model
model = sm.OLS(y, X).fit()

# Display model summary
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:                    BMI   R-squared:                       0.032
Model:                            OLS   Adj. R-squared:                  0.031
Method:                 Least Squares   F-statistic:                     33.13
Date:                Mon, 30 Mar 2026   Prob (F-statistic):           2.26e-33
Time:                        12:15:57   Log-Likelihood:                -16995.
No. Observations:                5001   AIC:                         3.400e+04
Df Residuals:                    4995   BIC:                         3.404e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept            29.5011      0.54

## Interpretation
* R² = 0.032 → The model explains only 3.2% of the variation in BMI
* Adjusted R² = 0.031 → Very similar, meaning predictors add little explanatory power
* F-statistic p-value = 2.26e-33 → The model is statistically significant overall

The model is statistically significant
But practically weak — sedentary time and other variables explain very little of BMI

* Coefficient = 0.0002
* p-value = 0.114
* Confidence interval crosses 0


Sedentary time is NOT statistically significant
It does not meaningfully predict BMI in this dataset
Even if significant, the effect size would be extremely small

Comparing the two regression models, both sedentary time and physical activity showed minimal ability to predict BMI. The sedentary time model explained slightly more variance (R² = 0.032) than the physical activity model (R² = 0.027), but the difference was negligible. In both models, the lifestyle variable of interest was not statistically significant, indicating weak associations with BMI. Diabetes diagnosis remained the only meaningful predictor, consistently associated with higher BMI. These findings suggest that BMI cannot be effectively predicted using single lifestyle variables and that additional clinical, behavioral, and socioeconomic factors are necessary.

# Ways the models can be improved:

The models could be improved by incorporating additional clinical and behavioral predictors. Variables such as:

* Waist circumference
* Fasting glucose
* HbA1c
* Blood pressure
* Dietary intake

These are strongly associated with BMI and diabetes risk and would likely increase R-squared values and improve statistical significance.

Addressing class imbalance through techniques such as **class weighting or oversampling could increase recall**, allowing the model to identify more true diabetes cases.

Additionally, reducing multicollinearity and adding interaction terms may stabilize coefficient estimates and improve model performance. Using non-**linear models such as random forests** may further enhance predictive accuracy by capturing complex relationships between lifestyle factors and health outcomes.